floris/tools/optimization/scipy/derive_downstream_turbines.py

```diff
@@ -0,0 +1,128 @@
+# Copyright 2021 NREL
+
+# Licensed under the Apache License, Version 2.0 (the "License"); you may not
+# use this file except in compliance with the License. You may obtain a copy of
+# the License at http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+# License for the specific language governing permissions and limitations under
+# the License.
+
+# See https://floris.readthedocs.io for documentation
+
+
+import matplotlib.pyplot as plt
+import numpy as np
+
+
+def derive_downstream_turbines(fi, wind_direction, wake_slope=0.10, plot_lines=False):
+    """Determine which turbines have no effect on other turbines in the
+    farm, i.e., which turbines have wakes that do not impact the other
+    turbines in the farm. This allows the user to exclude these turbines
+    from a control setpoint optimization, for example. This function
+    assumes a very simplified wake function where the wakes are assumed
+    very wide, with an initial width of twice the rotor diameter. This wake
+    model is defined to be very conservative and provide unreasonably wide
+    wakes to ensure all conditions, including variability and high turbulence,
```

Comment:
what would be considered high turbulence? Like, would this default wake model still be conservative for TI = 15% or 20%?

Reply:
Really good points, Eric! I had not looked at this with too much detail before but it does have a notable impact on the results, surely. Here is a comparison of the simple wake model with `wake_slope=0.10` against the default FLORIS model with `turbulence_intensity=0.20`:
![floris_vs_simple_model](https://user-images.githubusercontent.com/22119448/130092341-5253db42-a2b4-4e40-962b-a3d177cb3c79.PNG)

The gray lines are the points in FLORIS where the flow starts falling below 99.5% of the freestream wind speed. Actually, the simple wake model is very comparable in wake locations with FLORIS at `turbulence_intensity=0.20`. Now, if we were to have `turbulence_intensity=0.30`, the simple wake model actually predicts narrower wakes. A more fitting `wake_slope` in the simple model would be `0.15`:
![floris_vs_simple_model_ti0p30](https://user-images.githubusercontent.com/22119448/130092856-6b95050d-b6ba-489b-87cd-b7a4e23d0516.PNG)

And by hand, `turbulence_intensity=0.40` seems well captured by `wake_slope=0.20`. And `turbulence_intensity=0.05` seems well captured by `wake_slope=0.025`:
![floris_vs_simple_model_ti0p05](https://user-images.githubusercontent.com/22119448/130093284-4fb9bc3e-955a-47eb-b4d5-00c4e6e3e883.PNG)

You can try this yourself with
```
# Copyright 2021 NREL

# Licensed under the Apache License, Version 2.0 (the "License"); you may not
# use this file except in compliance with the License. You may obtain a copy of
# the License at http://www.apache.org/licenses/LICENSE-2.0

# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
# License for the specific language governing permissions and limitations under
# the License.

# See https://floris.readthedocs.io for documentation


import os

import matplotlib.pyplot as plt
import numpy as np
import floris.tools as wfct
from floris.tools.optimization.scipy.yaw import YawOptimization
from floris.tools.optimization.scipy.derive_downstream_turbines import derive_downstream_turbines


print("Running FLORIS with no yaw...")
# Instantiate the FLORIS object
file_dir = os.path.dirname(os.path.abspath(__file__))
fi = wfct.floris_interface.FlorisInterface(
    os.path.join(file_dir, "../../../example_input.json")
)

# Set turbine locations to 3 turbines in a row
D = fi.floris.farm.turbines[0].rotor_diameter
layout_x = [0, 7 * D, 14 * D, 0, 7 * D, 14 * D]
layout_y = [0, 0, 0, 7 * D, 7 * D, 7 * D]
fi.reinitialize_flow_field(layout_array=(layout_x, layout_y))
fi.calculate_wake()

# Initial power output
power_initial = fi.get_farm_power()



plot_lines = True
wake_slope = 0.10
fi.reinitialize_flow_field(wind_direction=267.)

ti=0.05
for wd in [270.]:
    fig, ax = plt.subplots(nrows=2)
    print('Wind direction: %.3f' % wd)  
    downstream_turbines = derive_downstream_turbines(fi=fi, wind_direction=wd, wake_slope=0.025, plot_lines=True, ax=ax[0])
    fi.reinitialize_flow_field(wind_direction=wd, turbulence_intensity=ti)
    hor_plane = fi.get_hor_plane()
    wake_edges = hor_plane.df[(hor_plane.df.u > 0.994*9.0) & (hor_plane.df.u < 0.995 * 9.0)].reset_index()
    wfct.visualization.visualize_cut_plane(hor_plane, ax=ax[1])
    ax[0].plot(wake_edges.x1, wake_edges.x2, '.', color='gray')
    ax[1].plot(wake_edges.x1, wake_edges.x2, '.', color='gray')
    for axi in ax:
        axi.axis('equal')
        axi.set_ylim([-500.0, 1500.0])
    plt.show()
```
where I added `ax` as an input to `derive_downstream_turbines()`.

Based on my findings, a good rule of thumb seems to be setting `wake_slope = 0.5 * turbulence_intensity`. If one were aiming to be conservative, a good approximation would be `wake_slope = turbulence_intensity`. For now, I will set `wake_slope = 0.30` as the default, which should be conservative and valid in realistically almost all relevant conditions. Thanks Eric!!

Reply:
Hey Bart, that's some really nice analysis. Thanks for looking into that! Do you mind just updating the documentation with a short statement about that rule of thumb (e.g., setting wake_slope = 0.5 * turbulence_intensity is a good rule of thumb when using the Gaussian wake model)?

Reply:
Done!